### DISCUSSION TITLE
Suggestion reguarding "contrib" folder in SourceForge download ("fixedpoint.zip")

### DISCUSSION THREAD

**Discussion by Andrew900460** [cc65/cc65#1700] [Ideas]
I hope this is not the wrong place to post this, I could find no better place to bring this up.

I just recently noticed the contributions folder on Source Forge, containing a bunch of interesting libraries (made by various people). The one that caught my eye was "fixedpoint.zip"; a fixed point library by Steffen Goerzig. Because I've recently (over the past few months) been trying to do fixed point math, so I can do some interesting physics stuff (on the NES).

So I ended up writing my own functions for doing fixed math, and also writing a function to convert an 8.8 fixed into a string. And then I also made an elaborate (but useful) system for printing my fixed point values to an *emulator external console*. FCEUX has Lua support, so I used that to interop with the 6502 and print messages to a console outside of the NES.

But digressions aside. I looked at the code in fixedpoint.zip and I noticed there was some incomplete code for multiplying "doublefixed" values. Understandably because math with 4-byte values sometimes requires another 4-byte "overflow" for certain reasons. And cc65 doesn't really support 8-byte values.

In his code, he wrote the comments: _"biggest cc65 data type has 4 bytes...so current macros for doublefixed...div and mul do not work properly"_

I saw this knowing that I actually recently made a function that can multiply 4-byte values, but it also retains the 4-byte overflow. And I felt like coming forward and helping by sharing my code to potentially fix the issue that Steffen Goerzig was having.

I would try to email Steffen Goerzig directly, but I did not see his email anywhere.
And I felt like this is an opportunity I should not pass up, and helping fix an unresolved issue in Goerzig's library.

So I hope I can be of help in this case :)

**Comment by polluks** [comment]
Interesting. In the past @oliverschmidt was managing contrib.

**Comment by oliverschmidt** [comment]
* I have no personal notes whatsoever regarding the _contrib_ content.
* I'm not managing _contrib_ anymore.

**Comment by mrdudz** [comment]
To me this "contrib" dir was always a bit of a strange dumping ground totally detached from cc65 itself. IMHO it would make sense to move all of it into a github repo (which may or may not be managed by the cc65 "organisation"), extract all those zips into seperate directories in the process, and perhaps even have a toplevel makefile that allows to check easily if that stuff even still builds (i expect a bunch of it broke by now).
I have no intentions to put work into this at this point however, except replacing existing .zip files with newer ones, should their author request it.

**Reply by Andrew900460** [reply]
Ok, I understand.
I noticed the folder while looking at the cc65 website, and was curious what the "contributions" link was about. So I dug into it and was surprised to see the fixedpoint.zip there

Like I said, I decided to take on a personal project where I wanted to make a particular simulation that could run on the NES, this started maybe a couple of months ago. So I knew I needed some sort of data type that could handle fractional data for object positions. And I knew historically, fixed point was sometimes used. So I had to write my own code for it.

When I noticed the fixedpoint.zip files, I was both amazed and *mildly infuriated* that I only noticed it the other day. Because that was the kind of thing that I would've wanted to find in a google search, before making my own code.

And in defense of the original creator of the library, it's a lot more "ironed out" compared to my code. I only have code for 8.8 fixed.
And he has plenty of functions for that, and even functions for 16.16 fixed, which is very useful if you need more precision and range.

So I would say... If anything, a library like that should be more "front and center" to cc65. It's a VERY general-purpose utility that greatly expands the number of things you can do with the 6502. I really wish I knew about that library earlier. Like it would be nice if in some way some of these libraries can be shown on the website in a more "you might find this useful" kind of fashion.
Like: "Click here for a list of downloads to some useful community made libraries"
Or something like that.

But yea, I understand you're probably busy with other aspects of cc65. I did say I would share some of the code that I think could round off the fixedpoint.zip code. But it also sounds like that code has kinda been long and forgotten. And it might be better if *someone else* continued to maintain it outside of cc65. Which probably makes sense.

But I will attach the code I was referring to up above ^ if you're curious.

But thank you for responding, I hope I am not taking up too much of your time here.
I'm really glad that a tool like cc65 exists, so I can experience what it might've been like to code during the 80's ;)

[umul32_with_overflow.zip](https://github.com/cc65/cc65/files/8218644/umul32_with_overflow.zip)

**Reply by Andrew900460** [reply]
To add a final remark to what I've said. It sounded like there might be a chance that this "contrib" folder might even be added to the git repo itself. If that is the case, that would allow me to fork it and make changes to that fixedpoint.h file directly. If I wanted to. And then I can make a pull request so It can be merged at your discretion.

So you don't have to put much time into it.

**Comment by mrdudz** [comment]
Sure. It could also automatically run tests to ensure everything still works with the current version of the compiler.

Right now this isnt something _i_ am going to spend time on. If anyone else wants to do it - go for it :)
